cxx/emissions/simple_string.hh

```diff
@@ -0,0 +1,191 @@
+    if (clean.length() < dirty.length()) {
+      // Probably an insertion.
+      d ? insertion.incorporate(1) : insertion.unincorporate(1);
+      corporate(clean, dirty.substr(1, std::string::npos), d);
+      return;
+    }
+
+    if (clean.length() > dirty.length()) {
+      // Probably a deletion.
+      d ? deletion.incorporate(1) : deletion.unincorporate(1);
+      corporate(clean.substr(1, std::string::npos), dirty, d);
+      return;
+    }
+
+    // Probably a substitution.
+    d ? substitution.incorporate(1) : substitution.unincorporate(1);
+    corporate(clean.substr(1, std::string::npos),
+              dirty.substr(1, std::string::npos),
+              d);
+  }
+
+  double logp(const std::pair<std::string, std::string>& x) const {
+    const_cast<SimpleStringEmission*>(this)->incorporate(x);
+    double lp = logp_score();
+    const_cast<SimpleStringEmission*>(this)->unincorporate(x);
+    return lp - logp_score();
+  }
+
+  double logp_score() const {
+    return substitution.logp_score() + insertion.logp_score() +
+        deletion.logp_score();
+  }
+
+  void transition_hyperparameters() {
+    substitution.transition_hyperparameters();
+    insertion.transition_hyperparameters();
+    deletion.transition_hyperparameters();
+  }
+
+  char random_character(std::mt19937* prng) {
+    std::uniform_int_distribution<char> uid(' ', '~' + 1);
+    return uid(*prng);
+  }
+
+  std::string sample_corrupted(const std::string& clean, std::mt19937* prng) {
+    substitution.prng = prng;
+    insertion.prng = prng;
+    deletion.prng = prng;
+    std::string s;
+    for (size_t i = 0; i < clean.length(); ++i) {
+      while (insertion.sample()) {
+        s += random_character(prng);
+      }
+      if (deletion.sample()) {
+        continue;
+      }
+      if (substitution.sample()) {
+        s += random_character(prng);
+      } else {
+        s += clean[i];
+      }
+    }
+    while (insertion.sample()) {
+      s += random_character(prng);
+    }
+    return s;
+  }
+
+  std::string propose_clean(
+      const std::vector<std::string>& corrupted,
+      std::mt19937* unused_prng) {
+    std::string clean;
+    // This implemention does simple voting per absolute string position.
+    // A better version would first average the corrupted string lengths to
+    // get a target length for clean, and then for each position i in
+    // clean, find the mode among the
+    // corrupted[j][i * corrupted[j].length / clean_length]
+    size_t i = 0;
+    while (true) {
+      std::unordered_map<char, int> counts;
+      int max_count = 0;
+      char mode;
+      for (const std::string& s: corrupted) {
+        char c;
+        if (i < s.length()) {
+          c = s[i];
+        } else {
+          c = '\0';
+        }
+        ++counts[c];
+        if (counts[c] > max_count) {
+          max_count = counts[c];
+          mode = c;
+        }
+      }
+      if (mode == '\0') {
+        return clean;
+      }
+      clean = clean + mode;
```

Comment:
Probably inconsequential but IMO might as well use `+=` since it calls `append` instead of copying the string (https://stackoverflow.com/questions/611263/efficient-string-concatenation-in-c/30401747#30401747)

Reply:
Done.